src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,302 @@
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private querier: QuerierClient;
+	private pusher: PusherClient;
+	private ready: boolean;
+	private tailCalls: Map<string, grpc.ClientReadableStream<TailResponse>>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.querier = new QuerierClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.pusher = new PusherClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.ready = false;
+		this.tailCalls = new Map();
+	}
+
+	public get available(): boolean {
+		return !this.ready;
+	}
+
+	/**
+	 *
+	 * Return $count of logs matching device_id in descending (BACKWARD) order.
+	 *
+	 * The logs are sorted by timestamp since Loki returns a distinct stream for each label combination.
+	 *
+	 * @param ctx
+	 * @param count
+	 */
+	public history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		return new Promise((resolve, reject) => {
+			const oneHourAgo = new Date(Date.now() - 10000 * 60);
+
+			const queryRequest = new QueryRequest();
+			queryRequest.setSelector(this.getDeviceQuery(ctx));
+			queryRequest.setLimit(Number.isFinite(count) ? count : 1000);
+			queryRequest.setStart(createTimestampFromDate(oneHourAgo));
+			queryRequest.setEnd(createTimestampFromDate());
+			queryRequest.setDirection(Direction.BACKWARD);
+
+			const deviceLogs: DeviceLog[] = [];
+			const call = this.querier.query(
+				queryRequest,
+				createOrgIdMetadata(String(ctx.belongs_to__application)),
+			);
+			call.on('data', (queryResponse: QueryResponse) => {
+				deviceLogs.push(
+					...this.fromStreamsToDeviceLogs(queryResponse.getStreamsList()),
+				);
+			});
+			call.on('error', (error: Error & { details: string }) => {
+				reject(new BadRequestError(`Query failed: ${error.details}"`));
+			});
+			call.on('end', () => {
+				resolve(_.orderBy(deviceLogs, 'timestamp', 'desc'));
+			});
+		});
+	}
+
+	public async publish(
+		ctx: LogWriteContext,
+		logs: Array<DeviceLog & { version?: number }>,
+	): Promise<any> {
+		const streams = this.fromDeviceLogsToStreams(ctx, logs);
+		try {
+			let tries = 0;
+			let nextBackoff = MIN_BACKOFF;
+			let prevBackoff = MIN_BACKOFF;
+			while (nextBackoff <= MAX_BACKOFF) {
+				try {
+					tries += 1;
+					return await this.push(ctx.belongs_to__application, streams);
+				} catch (error) {
+					// only retry when Loki is down or rate limiting
+					if (
+						tries <= MAX_RETRIES &&
+						[grpc.status.UNAVAILABLE, grpc.status.RESOURCE_EXHAUSTED].includes(
+							error.code,
+						)
+					) {
+						this.ready = false;
+						await sleep(nextBackoff);
+						// fibonacci backoff
+						nextBackoff = nextBackoff + prevBackoff;
+						prevBackoff = nextBackoff - prevBackoff;
+					} else {
+						throw new BadRequestError(`Publish failed: ${error.details}`);
+					}
+				}
+			}
+		} finally {
+			this.ready = true;
+		}
```

Comment:
Since this is in a `finally` it means that if we run out of retries we'll mark as ready even though we're not - that probably needs fixing

Reply:
The backend will retry publishing and mark the backend as not ready, after the max backoff is reached it throws and the logs are essentially dropped. The backend is marked `ready` again so that the next batch can attempt to publish. If its not marked as ready here, not sure how else it would be.

Reply:
I've removed the logic that sets ready to false when retrying. Not convinced its a good solution, will test if back-off is enough.